python/lance/_lib.pyx

```diff
@@ -273,28 +276,43 @@ cdef class FileSystemDataset(Dataset):
         c_version = GetResultValue(self.lance_dataset.latest_version())
         return _dataset_version_to_json(c_version)
 
-    def append_column(self, field: Field, func: Callable[[pyarrow.Table], pyarrow.Array]) -> FileSystemDataset:
+    def _append_column_expr(self, field: Field, expression: Expression) -> FileSystemDataset:
+        cdef CExpression c_expression = expression.unwrap()
+        cdef shared_ptr[CField] c_field = pyarrow_unwrap_field(field)
+        cdef shared_ptr[CLanceDataset] dataset = GetResultValue(self.lance_dataset.AddColumn(c_field, c_expression))
+        return FileSystemDataset.wrap(static_pointer_cast[CDataset, CLanceDataset](dataset))
+
+    def append_column(
+            self,
+            field: Field,
+            value: Union[Callable[[pyarrow.Table], pyarrow.Array], Expression],
+    ) -> FileSystemDataset:
         """Append a new column.
 
         Parameters
         ----------
         field : pyarrow.Field
             The name and schema of the newly added column.
-        func : Callback[[pyarrow.Table], pyarrow.Array]
+        value : Callback[[pyarrow.Table], pyarrow.Array], pyarrow.compute.Expression
```

Comment:
Callable is not tested?

Reply:
As mentioned above, callable / function was merged from previous PR, w/ their tests.

https://github.com/eto-ai/lance/blob/fd73142c2a1b8a165d2fce1e0da2c68372ded7cd/python/lance/tests/test_schema_evolution.py#L22